### DISCUSSION TITLE
StreamedResponse / can't get response client side (sourceEvents.readyState = 0)

### DISCUSSION THREAD

**Discussion by thgauthier** [symfony/symfony#49825] [Q&A] [answered]
hi,
I have followed symfony doc [streaming-a-response](https://symfony.com/doc/current/components/http_foundation.html#streaming-a-response)
Using symfony server (symfony serve)
But **first error**: EventSource's response has a MIME type ("text/html") that is not "text/event-stream". Aborting the connection.
Solved by adding : 
`$response->headers->set('Content-Type', 'text/event-stream');`
**Second error** : Failed to start the session because headers have already been sent
Solved by removing : `$response->send();` and adding `return $response;` thank you @stefangr,"_**The referenced documentation shows the standalone case (were the framework is not used).**_"
**Third error**: client side error with sourceEvents.readyState = 0 (0 = connecting) but status code = 200 (ok) at first attempt.
How may I success to get the console.log result keeping connection open for at least 4 seconds in that case ?

![Capture d’écran du 2023-03-27 09-33-51](https://user-images.githubusercontent.com/95478233/227872227-d0692686-ab67-48bb-9f99-65c48368ac48.png)
![Capture d’écran du 2023-03-27 09-33-58](https://user-images.githubusercontent.com/95478233/227872249-72ec8f06-0196-49d1-b7aa-4b3f55a49709.png)
![Capture d’écran du 2023-03-27 09-35-35](https://user-images.githubusercontent.com/95478233/227872558-c83332d9-9351-4365-8cf7-3be933f59d0d.png)
Thanks in advance for your answers
Thierry

**Comment by stefangr** [comment]
I looked into the [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) and [server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events).

The last link refers to an [example in PHP](https://github.com/mdn/dom-examples/blob/main/server-sent-events/sse.php).

From that code I can only conclude the syntax of the response is not correct.

**Reply by thgauthier** [reply]
you are right !
I can get something working with a syntax of the response like : `echo 'data: String message what ever needed.... ',"\n\n";`
![Capture d’écran du 2023-03-28 06-07-11](https://user-images.githubusercontent.com/95478233/228127463-a478c73c-4362-4d61-a966-4f956f21e103.png)
![Capture d’écran du 2023-03-28 06-25-48](https://user-images.githubusercontent.com/95478233/228128046-a189a8cd-16ec-4b5c-a869-3a08c603330d.png)

But any `flush();` `ob_end_flush();` `ob_flush()` `ob_start()` method do something interesting to get a result like [Symfony doc](https://symfony.com/doc/current/components/http_foundation.html#streaming-a-response) let me expect.

I might messing something.

What is also very strange each console.log message are display every 3 seconds.

In fact what I need to do is to start a bash process from a controller and send the output to the client side until the process stop.

As far as I understand the callback function run every 3 seconds so I can't run my bash process inside that callback... ?

**Reply by stefangr** [reply]
From https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events-intro

> Clients will reconnect if the connection is closed; a client can be told to stop reconnecting using the HTTP 204 No Content response code.

As you don't keep your connection open, the EventSource will reopen the connection.

**Reply by thgauthier** [reply]
Yes you're right and you may define the re-connection delay using the retry field in the [event stream format ](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format) : 
`echo "retry: 1000\ndata: This is a message at time " . date(DATE_COOKIE),"\n\n";`

**Comment by thgauthier** [comment, marked as answer]
So I found a solution working with apache2 and nginx **_but not with the symfony dev server_**
![code](https://user-images.githubusercontent.com/95478233/229267517-ef94f049-1730-4adc-8ed7-c7354a52a9b3.png)
`$response->headers->set('X-Accel-Buffering', 'no');` is only needed with nginx server with no effect if remained with apache2 server.
The only thing I don't know : How send a HTTP 204 response code inside the callBack function : 
`$response = new StreamedResponse();`
`$response->setCallback(function () use ($response) {`
`$response->setStatusCode(204);`
`return $response;` or `$response->send();` or both doesn't work.
As you may notice I send a stop string server side and if received by the client it close the connection. `sourceEvents.close();`
I you'd prefer sending a HTTP 204 response instead.
